File: JassToTs/Jass/Token.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Jass
{
    /// <summary> Виды токенов </summary>
    class TokenKind
    {
        #region виды токенов

        /// <summary> однострочный комментарий </summary>
        public const string lcom = "lcom";
        /// <summary> многострочный комментарий (в JASS отсутствуют)</summary>
        [Obsolete("в JASS только однострочные комментарии начинающиеся с //")]
        public const string mcom = "mcom";

        /// <summary> десятичное целое </summary>
        public const string ndec = "ndec";
        /// <summary> восьмеричное целое </summary>
        public const string oct = "oct";
        /// <summary> шеснадцатиричное целое в формате 0xNN </summary>
        public const string xhex = "xhex";
        /// <summary> шеснадцатиричное целое в формате $NN </summary>
        public const string dhex = "dhex";
        /// <summary> действительное </summary>
        public const string real = "real";
        /// <summary> шеснадцатиричное из 4х ASCII символов, записавыается в апостравах </summary>
        public const string adec = "adec";

        /// <summary> оператор </summary>
        public const string oper = "oper";

        /// <summary> строка в кавычках (в JASS единственный тип строк) </summary>
        public const string dstr = "dstr";
        /// <summary> строка в апострофах (, в JASS в апострафах хранятся целые, смотри <see cref="adec"/>) </summary>
        [Obsolete("не используется, в JASS единственный тип строк: заключённые в кавычках")]
        public const string sstr = "sstr";

        /// <summary> идентификатор </summary>
        public const string name = "name";

        /// <summary> базовый тип </summary>
        public const string btyp = "btyp";
        /// <summary> ключевое слово </summary>
        public const string kwd = "kwd";

        /// <summary> null значение </summary>
        public const string @null = "null";
        /// <summary> булевое значение </summary>
        public const string @bool = "bool";

        /// <summary> перевод строки </summary>
        public const string ln = "ln";

        /// <summary> левая скобка </summary>
        public const string lbra = "lbra";
        /// <summary> правая скобка </summary>
        public const string rbra = "rbra";
        /// <summary> левая квадратная скобка </summary>
        public const string lind = "lind";
        /// <summary> правая квадратная скобка </summary>
        public const string rind = "rind";

        /// <summary> макросы YDWE, пока что преобразуются в комментарий </summary>
        public const string ymacr = "ymacr";

        #endregion

        /// <summary> связь между видом и типом </summary>
        static Dictionary<string, string> TypeByKind = new Dictionary<string, string>
        {
            { lcom,  TokenType.comm },
            { mcom,  TokenType.comm },
            { ndec,  TokenType.val },
            { oct,   TokenType.val },
            { xhex,  TokenType.val },
            { dhex,  TokenType.val },
            { real,  TokenType.val },
            { adec,  TokenType.val },
            { oper,  TokenType.oper },
            { dstr,  TokenType.val },
            { sstr,  TokenType.val },
            { name,  TokenType.name },
            { btyp,  TokenType.name },
            { kwd,   TokenType.kwd },
            { @null, TokenType.val },
            { @bool, TokenType.val },
            { ln,    TokenType.br },
            { lbra,  TokenType.par },
            { rbra,  TokenType.par },
            { lind,  TokenType.par },
            { rind,  TokenType.par },
            // макросы считаем за коментарии
            { ymacr, TokenType.comm },
        };

        /// <summary> получить тип </summary>
        /// <param name="kind"> вид токена </param>
        public static string GetType(string kind) => TypeByKind[kind];
    }

    /// <summary> Типы токенов </summary>
    class TokenType
    {
        /// <summary> разделитель </summary>
        public const string br = "br";
        /// <summary> значение (безымянная константа) </summary>
        public const string val = "val";
        /// <summary> ключевое слово </summary>
        public const string kwd = "kwd";
        /// <summary> идентификатор </summary>
        public const string name = "name";
        /// <summary> оператор </summary>
        public const string oper = "oper";
        /// <summary> скобки </summary>
        public const string par = "par";
        /// <summary> комментарий </summary>
        public const string comm = "comm";
    }

    /// <summary> Токен </summary>
    class Token
    {
        public string Type => TokenKind.GetType(Kind);
        public string Kind = "";
        public int Line = 0;
        public int Col = 0;
        public int Pos = 0;
        public string Text = "";
        public override string ToString() => $"{Line},{Col} [{Type}|{Kind}]: {Text}";
        public Token Clone() => new Token { Kind = Kind, Line = Line, Col = Col, Pos = Pos, Text = Text };
    }
}

File: JassToTsPython/Token.py
import enum

# Виды токенов
class TokenKind(enum.Enum):
    # однострочный комментарий
    lcom = "lcom"
    # многострочный комментарий (в JASS отсутствуют)</summary>
    #[Obsolete("в JASS только однострочные комментарии начинающиеся с //")]
    mcom = "mcom"

    # десятичное целое
    ndec = "ndec"
    # восьмеричное целое
    oct = "oct"
    # шеснадцатиричное целое в формате 0xNN
    xhex = "xhex"
    # шеснадцатиричное целое в формате $NN
    dhex = "dhex"
    # действительное
    real = "real"
    # шеснадцатиричное из 4х ASCII символов, записавыается в апостравах
    adec = "adec"

    # оператор
    oper = "oper"

    # строка в кавычках (в JASS единственный тип строк)
    dstr = "dstr"
    # строка в апострофах (, в JASS в апострафах хранятся целые, смотри <see cref="adec"/>)
    #[Obsolete("не используется, в JASS единственный тип строк: заключённые в кавычках")]
    sstr = "sstr"

    # идентификатор
    name = "name"

    # базовый тип
    btyp = "btyp"
    # ключевое слово
    kwd = "kwd"

    # null значение
    null = "null"
    # булевое значение
    bool = "bool"

    # перевод строки
    ln = "ln"

    # левая скобка
    lbra = "lbra"
    # правая скобка
    rbra = "rbra"
    # левая квадратная скобка
    lind = "lind"
    # правая квадратная скобка
    rind = "rind"

    # макросы YDWE, пока что преобразуются в комментарий
    ymacr = "ymacr"

class TokenType(enum.Enum):
    # разделитель
    br = "br"
    # значение (безымянная константа)
    val = "val"
    # ключевое слово
    kwd = "kwd"
    # идентификатор
    name = "name"
    # оператор
    oper = "oper"
    # скобки
    par = "par"
    # комментарий
    comm = "comm"

# связь между видом и типом
TypeByKind = {
    TokenKind.lcom:  TokenType.comm,
    TokenKind.mcom:  TokenType.comm,
    TokenKind.ndec:  TokenType.val,
    TokenKind.oct:   TokenType.val,
    TokenKind.xhex:  TokenType.val,
    TokenKind.dhex:  TokenType.val,
    TokenKind.real:  TokenType.val,
    TokenKind.adec:  TokenType.val,
    TokenKind.oper:  TokenType.oper,
    TokenKind.dstr:  TokenType.val,
    TokenKind.sstr:  TokenType.val,
    TokenKind.name:  TokenType.name,
    TokenKind.btyp:  TokenType.name,
    TokenKind.kwd:   TokenType.kwd,
    TokenKind.null:  TokenType.val,
    TokenKind.bool:  TokenType.val,
    TokenKind.ln:    TokenType.br,
    TokenKind.lbra:  TokenType.par,
    TokenKind.rbra:  TokenType.par,
    TokenKind.lind:  TokenType.par,
    TokenKind.rind:  TokenType.par,
    # макросы считаем за коментарии
    TokenKind.ymacr: TokenType.comm
}

# получить тип
# <param name="kind"> вид токена </param>
def GetType(kind): 
    return TypeByKind[kind]

class Token:
    def __init__(self, col = 0, line = 0, pos = 0, text = "", kind = ""):
        self.Col = col
        self.Line = line
        self.Pos = pos
        self.Text = text
        self.Kind = kind

    def getType(self):
        return GetType(self.Kind)

    def __str__(self):
        return "{},{} [{}|{}]: {}".format(self.Line, self.Col, self.getType(), self.Kind, self.Text)

    def Clone(self):
        return Token(self.Col, self.Line, self.Pos, self.Text, self.Kind)
